Restructure building-from-source/getting-started page.

ScottTodd opened the pull request:
This restructures the main "how to build from source" page that we have, highlighting how simple "configure, build, test" _can_ be then going into what settings/scripts we recommend and why.

### Hosted pages

| | |
| --- | --- |
| Current page | https://openxla.github.io/iree/building-from-source/getting-started/ | 
| Preview this PR | https://scotttodd.github.io/iree/building-from-source/getting-started/ |

### Summary of changes:

* Add basic configure -> build _then_ explain settings
* Add TODOs for Docker / dev containers and CMake presets
  * https://github.com/openxla/iree/pull/12921
  * https://github.com/openxla/iree/issues/5804
* Discuss optional components (to convey "CUDA is enabled by default")
* Add section on running samples like `hello_world_embedded`
* Refine language throughout

ScottTodd committed:
Restructure building-from-source/getting-started page.

* Add basic configure -> build _then_ explain settings
* Add TODOs for Docker / dev containers and CMake presets
* Discuss optional components (to convey "CUDA is enabled by default")
* Add section on running samples like `hello_world_embedded`
* Refine language throughout

ScottTodd commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
     See also our [developer documentation for ccache](https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/ccache.md).
 
-## What's next?
+### Optional components
+
+By default, the CMake build includes:
+
+* All compiler targets (`llvm-cpu`, `cuda`, `vulkan-spirv`, etc.)
+* All runtime HAL drivers (`local-task`, `cuda`, `vulkan`, etc.)
+* All compiler input formats (StableHLO, TOSA, etc.)
+* All compiler output formats (VM bytecode, C)
```
I'm not sure about including this full section here... it's pretty large when scanning through the page. What do others think?

I could also experiment with formatting (e.g. table or task list instead of list).

ScottTodd commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
+These can be changed via the `IREE_` CMake options listed in the root
+[`CMakeLists.txt`](https://github.com/openxla/iree/blob/main/CMakeLists.txt).
+
+### Extensions and project integrations
+
+When using IREE within other projects, you can register compiler plugins and
+runtime HAL drivers. You can also bring your own copy of LLVM and some other
+tools. See the root
+[`CMakeLists.txt`](https://github.com/openxla/iree/blob/main/CMakeLists.txt)
+for details.
```
Also not sure about this new section. Could be useful to a few people to have this advertised up front though.

pcoet replied:
I like this testing section and the "Optional components" section above. It's useful info, and the section titles and ordering make it clear that it's not necessary for building.

ScottTodd commented on docs/website/docs/building-from-source/getting-started.md:
````diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
+cmake --build ../iree-build --target iree-run-tests
+```
+
+To run only certain tests, we have a
+[helper script](https://github.com/openxla/iree/blob/main/build_tools/cmake/ctest_all.sh)
+that converts environment variables into ctest filters:
````
Should we mention other scripts like `build_all.sh` too?

GMNGeoffrey replied:
These scripts are really designed for CI (and say so at the top). They were made possible to use locally for the purpose of keeping CI reproducible, not for providing an ergonomic entry point for developers. ctest_all.sh is mostly pretty usable as a development tool, but especially the build scripts were really not aimed at that. I'd prefer we not direct users to them and create expectations that they are something other than what they are.

pcoet commented on docs/website/docs/building-from-source/getting-started.md:
````diff
@@ -70,66 +67,75 @@ cd iree
+```
+
+!!! Caution "Caution - slow builds"
+    The compiler build is complex. You will want a powerful machine and to tune
````
If you want, you could give some idea of the expected build time. Depending on the context, "slow" could mean 12 minutes or 12 hours.

stellaraccident replied:
"As of March, 2023, expect that a full compiler build on a 64 core Linux machine with good I/O may take 5-7 minutes. On MacOS, building non-universal on an M1 Mini should take around 15 minutes."

ScottTodd replied:
I'm hesitant to add text but can wordsmith a bit. On low spec machines (2 core github actions runners), we've seen builds take 4-6h+

pcoet commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
+    This has two limitations:
+
+    1. Large tests are excluded from the build by default
+    2. Some tests require hardware like a GPU and will fail on unsupported systems
```
If I run `ctest --test-dir ../iree-build/` on my Mac or gLinux, will it just work? It would be nice if we had a test option that would exclude tests that aren't supported by the hardware (I realize that might not be feasible).

pcoet replied:
Never mind. I just saw the section about the helper script that uses env vars to create filters.

ScottTodd replied:
> It would be nice if we had a test option that would exclude tests that aren't supported by the hardware (I realize that might not be feasible).

Discussion with some ideas at https://github.com/openxla/iree/issues/12042

pcoet reviewed (approved):
LGTM. These changes make the doc clearer, more prescriptive, and  more usable.

stellaraccident commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -70,66 +67,75 @@ cd iree
-    # Recommended for simple development using clang and lld:
-    cmake -GNinja -B ../iree-build/ -S . \
+    # Recommended development options using clang and lld:
+    cmake -G Ninja -B ../iree-build/ -S . \
```
If recommending a debug build, we really should be directing Linux users to include:

```
-DIREE_ENABLE_SPLIT_DWARF=ON
-DIREE_ENABLE_THIN_ARCHIVES=ON
```

Especially for more limited machines, these have massive space/overhead savings when debug symbols are present.

GMNGeoffrey replied:
What is the disadvantage to turning those on by default?

stellaraccident replied:
They really only should be for "dev" use as they make the artifacts non-relocatable and such.

stellaraccident commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
-<!-- TODO(scotttodd): "at this point you can..." -->
+The default build does _not_ include:
+
+* Compiler or runtime bindings (Python, TFLite, etc.)
```
I think that (perhaps in a followup), we should move the Python configuration to this page as an optional component. We can leave the other random integrations on an aux page, but Python support is a core supported features and should be here.

ScottTodd replied:
I'd be curious how many developers/contributors actually need to build and edit the Python bindings from source. Probably more over time as JAX / PJRT integration picks up speed.

The environment setup steps on https://openxla.github.io/iree/building-from-source/python-bindings-and-importers/ are fairly complicated for casual users that want to follow quickstart instructions. Maybe the dev container can help there (configure the right version of Python by default).

stellaraccident replied:
I don't know/care - I just know that I am tired of flipping to another page for a perfectly supported/normal part of the build. Most of our current users do in fact use it through the Python bindings, afaik.

ScottTodd replied:
Okay, follow-up SGTM. I want to see how the dev container lands to see if we can recommend that for Python instead of having the notes about checking for python2/python3 and installing the extra prerequisites. Another (less-invasive) idea is to make the "Environment setup" section collapsible

GMNGeoffrey commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -50,16 +46,17 @@ python package [requests](https://requests.readthedocs.io/en/latest/) to run.
         for details.
 
-## Clone and build
+<!-- TODO(scotttodd): add notes about Docker and/or dev containers  -->
```
Link issues instead?

ScottTodd commented on docs/website/docs/building-from-source/getting-started.md:
```diff
@@ -138,32 +144,102 @@ cmake --build ../iree-build/
+ls ../iree-build/install/bin/
+
+# Run a standalone sample application
+../iree-build/install/bin/hello_world_embedded
```
I was hoping to use `install` to make this section easier to follow, but Linux defaults to a system-wide install at `/usr/local/bin`:
```
$ cmake --build ../iree-build/ --target install
[0/2] Re-checking globbed directories...
[0/1] Install the project...
-- Install configuration: "Release"
-- Installing: /usr/local/bin/iree-flatcc-cli
```

I'll update to just use the non-install paths in the build directory (rather than add `-DCMAKE_INSTALL_PREFIX=../iree-build/install` to the flags above)

ScottTodd committed:
Use full paths instead of relying on `install` to run samples/tools.

`install` defaults to a system-wide install on Linux, which is not what most users will want for IREE samples, and I don't want to include `-DCMAKE_INSTALL_PREFIX` in all the sample commands

ScottTodd committed:
Address review comments.

* Trim MSVC note
* Link to issues/PRs in TODOs
* Add data point for "slow builds"
* Recommend `IREE_ENABLE_SPLIT_DWARF` and `IREE_ENABLE_THIN_ARCHIVES`
